lib/charms/haproxy/v0/haproxy_route.py

```diff
@@ -0,0 +1,543 @@
+        relation_name: str = HAPROXY_ROUTE_RELATION_NAME,
+    ) -> None:
+        """_summary_
+
+        Args:
+            charm (CharmBase): _description_
+            relation_name (str, optional): _description_. Defaults to HAPROXY_ROUTE_RELATION_NAME.
+        """
+        super().__init__(charm, relation_name)
+
+        self._relation_name = relation_name
+        self.charm = charm
+        on = self.charm.on
+        self.framework.observe(on[self._relation_name].relation_created, self._configure)
+        self.framework.observe(on[self._relation_name].relation_changed, self._configure)
+        self.framework.observe(on[self._relation_name].relation_broken, self._configure)
+
+    @property
+    def relations(self):
+        """The list of Relation instances associated with this endpoint."""
+        return list(self.charm.model.relations[self._relation_name])
+
+    def _configure(self, event: EventBase) -> None:
+        """_summary_
+
+        Args:
+            event (EventBase): _description_
+        """
+        try:
+            if relations := self.relations:
+                self.get_data(relations)
+                self.on.data_available.emit()  # type: ignore
+        except HaproxyRouteInvalidRelationDataError:
+            logger.exception("Invalid requirer data, skipping.")
+            return
+
+    def get_data(self, relations: list[Relation]) -> HaproxyRouteRequirersData:
+        """Fetch the remote (requirer) app and units' databags."""
+        haproxy_route_requirers_data: list[HaproxyRouteRequirerData] = []
+        for relation in relations:
+            try:
+                application_data = self._get_requirer_application_data(relation)
+                units_data = self._get_requirer_units_data(relation)
+                harpoxy_route_requirer_data = HaproxyRouteRequirerData(
+                    application_data=application_data, units_data=units_data
+                )
+                haproxy_route_requirers_data.append(harpoxy_route_requirer_data)
+            except DataValidationError as exc:
+                logger.exception("haproxy-route data validation failed for relation %s", relation)
+                raise HaproxyRouteInvalidRelationDataError(
+                    f"haproxy-route data validation failed for relation: {relation}"
+                ) from exc
+        return HaproxyRouteRequirersData(haproxy_route_requirers_data=haproxy_route_requirers_data)
+
+    def _get_requirer_units_data(self, relation: Relation) -> list[RequirerUnitData]:
+        """Fetch and validate the requirer's app databag."""
+        requirer_units_data: list[RequirerUnitData] = []
+
+        for unit in relation.units:
+            databag = relation.data[unit]
+            try:
+                data = RequirerUnitData.load(databag)
+                requirer_units_data.append(data)
+            except DataValidationError:
+                logger.exception("Invalid requirer application data for %s", unit)
```

Comment:
It seems there are many logger.exception for the same thing. Or could it be different depending on the point?

Reply:
I agree, we should logger.exception only once, the remaining might be logger.error only

Reply:
When the exception is reraised, you don't necessarily need to log it the exception has a meaningful message
When you raise a new exception from the previous, I'd log the message of the previous exception, and ensure a meaningful message is added to the new exception
When you reach the top, if you "ignore" the exception, log it using logger.exception, otherwise, the unhandled exceptions will anyway endup in the logs (so no need to log it again)

Reply:
Updated the logging to not use `logger.exception` everywhere